### ISSUE TITLE
Accessibility tool to use with CircleCI

### ISSUE THREAD

**Issue by master12** [ONRR/doi-extractives-data#4638]
Choose an accessibility tool to run an audit locally and during the circleci build

**Comment by brentryanjohnson** [comment]
In the past, we have manually audited for accessibility with two separate tools:

- Lighthouse (from Chrome dev tools)
- pa11y CLI

The latter has an [option for CI integration](https://github.com/pa11y/pa11y-ci). 
Here is a [list of other options](https://www.w3.org/WAI/ER/tools/) we could look at (not all are CI options, but some are listed).

**Comment by jlouis-onrr** [comment]
Im going to look into these options and see how they compare to pa11y CLI.  Right off the bat, I'm liking the pa11y CLI especially since you can define a sitemap.

**Comment by jlouis-onrr** [comment]
Here are my thoughts on the approach:

- Use the [Gatsby Sitemap Plugin](https://www.gatsbyjs.org/packages/gatsby-plugin-sitemap/?=sitemap) to Generate sitemap
- Use CircleCI, during build, run pa11y-ci against the generated sitemap to output any accessibility errors

Seems like a pretty straightforward approach.  Any thoughts?

**Comment by brentryanjohnson** [comment]
Seems to me like a sound approach to me, @jlouis-onrr. What would be the next step?

**Comment by jlouis-onrr** [comment]
@brentryanjohnson My next step is to configure the [Gatsby Sitemap plugin](https://www.gatsbyjs.org/packages/gatsby-plugin-sitemap/?=sitemap) to generate the sitemap.  Are there any exclusions that we want to leave out of the sitemap?

**Comment by jlouis-onrr** [comment]
@master12 I was able to get the CircleCi CLI tool running locally to perform tests against the build and here is what Im currently seeing get logged:

`$ circleci local execute --job build
Downloading latest CircleCI build agent...
Docker image digest: sha256:e8c08f3dc363f9b90915266df7220b9d4c1413fd66aed162a854da5ae1ae88c4
====>> Spin up Environment
Build-agent version 1.0.17801-1d37c01b (2019-10-15T22:27:37+0000)
Docker Engine Version: 19.03.2
Kernel Version: Linux 3d13c261d187 4.9.184-linuxkit #1 SMP Tue Jul 2 22:58:16 UTC 2019 x86_64 Linux
Starting container circleci/node:10.16.0
  image cache not found on this host, downloading circleci/node:10.16.0
10.16.0: Pulling from circleci/node
a4d8138d0f6b: Pulling fs layer
dbdc36973392: Pulling fs layer
f59d6d019dd5: Pulling fs layer
aaef3e026258: Pulling fs layer
6e454d3b6c28: Pulling fs layer
c717a7c205aa: Pulling fs layer
038b3cfb4595: Pulling fs layer
a6d4e452109c: Pulling fs layer
f2e9874fa393: Pulling fs layer
f03fb4b9009d: Pulling fs layer
1b0e654c8a66: Pulling fs layer
754326050c2b: Pulling fs layer
203be6c8dd18: Pulling fs layer
7fe13c337c1e: Pulling fs layer
639ca3ec0bb9: Pulling fs layer
a4890fe06cc8: Pulling fs layer
a6b2dd11493b: Pulling fs layer
4e0d7f305b34: Pulling fs layer
aaef3e026258: Waiting
6e454d3b6c28: Waiting
c717a7c205aa: Waiting
038b3cfb4595: Waiting
a6d4e452109c: Waiting
f03fb4b9009d: Waiting
1b0e654c8a66: Waiting
754326050c2b: Waiting
203be6c8dd18: Waiting
7fe13c337c1e: Waiting
639ca3ec0bb9: Waiting
a4890fe06cc8: Waiting
f2e9874fa393: Waiting
a6b2dd11493b: Waiting
4e0d7f305b34: Waiting
f59d6d019dd5: Verifying Checksum
f59d6d019dd5: Download complete
dbdc36973392: Verifying Checksum
dbdc36973392: Download complete
aaef3e026258: Download complete
c717a7c205aa: Verifying Checksum
c717a7c205aa: Download complete
a4d8138d0f6b: Verifying Checksum
a4d8138d0f6b: Download complete
a6d4e452109c: Verifying Checksum
a6d4e452109c: Download complete
f2e9874fa393: Verifying Checksum
f2e9874fa393: Download complete
f03fb4b9009d: Verifying Checksum
f03fb4b9009d: Download complete
1b0e654c8a66: Verifying Checksum
1b0e654c8a66: Download complete
a4d8138d0f6b: Pull complete
dbdc36973392: Pull complete
f59d6d019dd5: Pull complete
038b3cfb4595: Verifying Checksum
038b3cfb4595: Download complete
203be6c8dd18: Verifying Checksum
203be6c8dd18: Download complete
aaef3e026258: Pull complete
7fe13c337c1e: Verifying Checksum
7fe13c337c1e: Download complete
754326050c2b: Download complete
639ca3ec0bb9: Download complete
a4890fe06cc8: Download complete
4e0d7f305b34: Download complete
a6b2dd11493b: Verifying Checksum
a6b2dd11493b: Download complete
6e454d3b6c28: Verifying Checksum
6e454d3b6c28: Download complete
6e454d3b6c28: Pull complete
c717a7c205aa: Pull complete
038b3cfb4595: Pull complete
a6d4e452109c: Pull complete
f2e9874fa393: Pull complete
f03fb4b9009d: Pull complete
1b0e654c8a66: Pull complete
754326050c2b: Pull complete
203be6c8dd18: Pull complete
7fe13c337c1e: Pull complete
639ca3ec0bb9: Pull complete
a4890fe06cc8: Pull complete
a6b2dd11493b: Pull complete
4e0d7f305b34: Pull complete
Digest: sha256:48c26a4dcf96e83090e340e2b274d3f83499891d6d0dc54146124ea82f4fc754
Status: Downloaded newer image for circleci/node:10.16.0
  using image circleci/node@sha256:48c26a4dcf96e83090e340e2b274d3f83499891d6d0dc54146124ea82f4fc754

Using build environment variables
  BASH_ENV=/tmp/.bash_env-localbuild-1571182629
  CI=true
  CIRCLECI=true
  CIRCLE_BRANCH=4638-AccessibilityToolToUseWithCircleCI
  CIRCLE_BUILD_NUM=
  CIRCLE_JOB=build
  CIRCLE_NODE_INDEX=0
  CIRCLE_NODE_TOTAL=1
  CIRCLE_REPOSITORY_URL=git@github.com:ONRR/doi-extractives-data.git
  CIRCLE_SHA1=d1bfcba91a00756155da4a0cd76aa5cc6ad74997
  CIRCLE_SHELL_ENV=/tmp/.bash_env-localbuild-1571182629
  CIRCLE_WORKING_DIRECTORY=~/project


The redacted variables listed above will be masked in run step output

====>> Checkout code
  #!/bin/bash -eo pipefail
mkdir -p /home/circleci/project && cd /tmp/_circleci_local_build_repo && git ls-files | tar -T - -c | tar -x -C /home/circleci/project && cp -a /tmp/_circleci_local_build_repo/.git /home/circleci/project
====>> Restoring Cache
Error: 
Skipping cache - error checking storage: not supported

Step failed
====>> Install node dependencies
  #!/bin/bash -eo pipefail
npm install

> sharp@0.22.1 install /home/circleci/project/node_modules/sharp
> (node install/libvips && node install/dll-copy && prebuild-install) || (node-gyp rebuild && node install/dll-copy)

info sharp Downloading https://github.com/lovell/sharp-libvips/releases/download/v8.7.4/libvips-8.7.4-linux-x64.tar.gz

> puppeteer@1.20.0 install /home/circleci/project/node_modules/puppeteer
> node install.js

Downloading Chromium r686378 - 114 Mb [] 100% 0.0s 
Chromium downloaded to /home/circleci/project/node_modules/puppeteer/.local-chromium/linux-686378

> node-sass@4.12.0 install /home/circleci/project/node_modules/node-sass
> node scripts/install.js

Downloading binary from https://github.com/sass/node-sass/releases/download/v4.12.0/linux-x64-64_binding.node
Download complete
Binary saved to /home/circleci/project/node_modules/node-sass/vendor/linux-x64-64/binding.node
Caching binary to /home/circleci/.npm/node-sass/4.12.0/linux-x64-64_binding.node

> core-js@2.6.9 postinstall /home/circleci/project/node_modules/@babel/polyfill/node_modules/core-js
> node scripts/postinstall || echo "ignore"


> core-js@2.6.9 postinstall /home/circleci/project/node_modules/@gatsbyjs/relay-compiler/node_modules/core-js
> node scripts/postinstall || echo "ignore"


> core-js@2.6.9 postinstall /home/circleci/project/node_modules/babel-runtime/node_modules/core-js
> node scripts/postinstall || echo "ignore"


> core-js-pure@3.1.4 postinstall /home/circleci/project/node_modules/core-js-pure
> node scripts/postinstall || echo "ignore"


> core-js@2.6.9 postinstall /home/circleci/project/node_modules/gatsby/node_modules/core-js
> node scripts/postinstall || echo "ignore"


> core-js@2.6.9 postinstall /home/circleci/project/node_modules/relay-runtime/node_modules/core-js
> node scripts/postinstall || echo "ignore"


> gatsby-telemetry@1.1.7 postinstall /home/circleci/project/node_modules/gatsby-telemetry
> node src/postinstall.js

(node:29) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 SIGINT listeners added. Use emitter.setMaxListeners() to increase limit

> node-sass@4.12.0 postinstall /home/circleci/project/node_modules/node-sass
> node scripts/build.js

Binary found at /home/circleci/project/node_modules/node-sass/vendor/linux-x64-64/binding.node
Testing binary
Binary is fine
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 2207 packages from 1150 contributors and audited 909044 packages in 143.597s
found 1 high severity vulnerability
  run `npm audit fix` to fix them, or `npm audit` for details
====>> Saving Cache
Error: 
Skipping cache - error checking storage: not supported

Step failed
====>> Build gatsby
  #!/bin/bash -eo pipefail
npm run build

> NRRD@1.0.0 build /home/circleci/project
> NODE_OPTIONS=--max_old_space_size=8192 gatsby build --prefix-paths

BASEURL:  undefined
success open and validate gatsby-configs — 0.127
success load plugins — 2.742
success onPreInit — 0.036
success delete html and css files from previous builds — 0.029
success initialize cache — 0.043
success copy gatsby files — 0.180
success onPreBootstrap — 0.048
Segmentation fault
npm ERR! code ELIFECYCLE
npm ERR! errno 139
npm ERR! NRRD@1.0.0 build: `NODE_OPTIONS=--max_old_space_size=8192 gatsby build --prefix-paths`
npm ERR! Exit status 139
npm ERR! 
npm ERR! Failed at the NRRD@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/circleci/.npm/_logs/2019-10-15T23_45_45_543Z-debug.log
Error: Exited with code 139
Step failed
Error: runner failed (exited with 101)
Task failed
Error: task failed`

**Comment by jlouis-onrr** [comment]
@brentryanjohnson What should our audit baseline scores that we test against be for the following Audit Categories?

Audit Baseline Scores:
- Performance 
- Accessibility
- Best Practices
- SEO

**Comment by brentryanjohnson** [comment]
@jlouis-onrr Baselines or targets? 

I think we still need to answer some questions first, especially for performance. If you run the audit now, it's heavily influenced by the throttling setting. I think we need to know (from analytics, presumably) if we should throttle or not, based on the proportion of users' bandwidth constraints. Furthermore, I don't think we yet know how the move to a database will impact our performance number. I don't know if we should concern ourselves with that or not.

I think our goal for the others should be 100. Whether that is realistic or not, I'm unsure. As for baseline, whatever we're scoring on those now could be our baseline.

**Comment by jlouis-onrr** [comment]
Definitely more questions to answer.  Right now, the audits are drastically different between no throttling which has pretty good scores and 4G and 3G throttling which currently gets lower scores especially in the performance arena.  I think I'll start off with the desktop no throttling and we can then add in different variations and measurements moving forward as more questions get answered and or if there are more requirements.  The configuration options in Lighthouse CLI are practically endless.

**Comment by brentryanjohnson** [comment]
Sounds good, @jlouis-onrr. Thank you!
